### PR TITLE
"Upload profiling data" mode

### PR DESCRIPTION
**Pull request by pfilipko1**
New subcommand: upload-file

## Description
Addition of a new subcommnad that enables uploading external .col files to Granulate Performance Studio.

## Related Issue
https://github.com/Granulate/gprofiler/issues/524

## Motivation and Context
It adds new feature, that enables

## How Has This Been Tested?
It was tested manually, using sample .col data without metadata and container names in such scenarios: 

- right file, right path
- wrong file, right path
- wrong path

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] I have updated the relevant documentation.
- [ ] I have added tests for new logic.

### CODE REVIEW

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:515]: Don't set `default` - it will be `None` which makes more sense

File: README.md
Comment (Jongy) [inline README.md:116]: UX wise, I think `--upload-collapsed-file` should be enough (so `-u` needn't be given if that one's given).

Alternatively - enforce having `-u` given if `--upload-collapsed-file` is given

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:404]: I think these 2 should also be accepted in the file - so file creator can specify the timeframe. Let's define it as 2 fields in the first line of the collapsed (2 UTC timestamps) and try to load them from there. If not given - use a NULL timestamp (i.e UTC timestamp of the epoch)

**Comment by Jongy** [comment]
Not done reviewing

File: gprofiler/main.py
Comment (pfilipko1) [reply gprofiler/main.py:396]: Done, info in the README.md

File: gprofiler/merge.py
Comment (Jongy) [inline gprofiler/merge.py:351]: You should enrich the metadata here with the result of `get_current_metadata`. Values in `data` take precedence, so we need to deep merge `data` onto the resulting dict from `get_current_metadata`.

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:506]: Did you try adding it as a subcommand? I think it makes more sense UX wise, and also allows us to structure the code better. With the current way this is written, the preconditions for gProfiler run are still checked in the upload mode, despite there's no need for that (checking for root, checking for init PID NS, ...).

What I had in mind is - `./gprofiler upload-file` will be a subcommand accepting only the relevant parameters - let's begin with `token`, `service-name` and later if needed we can add others such as `server-upload-timeout`. It will reject other parameters because they are truly not used. `./gprofiler` will run the default behavior (i.e what we currently have) and accept all arguments.
See [this](https://stackoverflow.com/questions/6365601/default-sub-command-or-handling-no-sub-command-with-argparse) SO question with an example of how to do that with `argparse`, and then there's a way to define the argument `--token` once, and have it available both for the default parser (`./gprofiler`) and for the upload subparser (`./gprofiler upload-file`)

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:828]: If we change it to be a subcommand, this diff will be eliminated because we won't need to change anything here

**Review by Jongy — CHANGES_REQUESTED** [review]
Left some comments, we still need to change the design IMO, mostly with argparse

File: gprofiler/merge.py
Comment (Jongy) [reply gprofiler/merge.py:343]: It's not - I created https://github.com/Granulate/gprofiler/pull/556 to remove it, you can merge it to your branch meanwhile, or wait for it to be merged

File: README.md
Comment (Jongy) [inline README.md:115]: `--upload-collapsed-file` that's old, needs to be updated

File: gprofiler/merge.py
Comment (Jongy) [reply gprofiler/merge.py:355]: It can be removed now

File: gprofiler/merge.py
Comment (Jongy) [inline gprofiler/merge.py:367]: Let's move parsing to a utility function `parse_iso8601_timestamp`? Place it next to `get_iso8601_format_time`.

File: gprofiler/merge.py
Comment (Jongy) [inline gprofiler/merge.py:364]: Check if line begins with comment (i.e `#`, and then the json itself can be immediately after it `#{...}` or with a space `# {}`. I suggest you just `json.loads(line[1:])` and it'll work for both cases

File: gprofiler/merge.py
Comment (Jongy) [inline gprofiler/merge.py:373]: total_samples can be removed

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:894]: "else" what? It can't happen right? So instead
```suggestion
            assert args.subcommands == "upload-file"
```

File: gprofiler/merge.py
Comment (Jongy) [inline gprofiler/merge.py:336]: Typing please :) `Dict[str, Any]` suffices

Comment (Jongy) [inline gprofiler/merge.py:336]: This is fairly generic. Let's move to utils and keep with generic name `merge_dicts`? Also names like `source` and `dest` are more clear IMO, I liked them [here](https://stackoverflow.com/a/20666342/797390)

File: gprofiler/merge.py
Comment (Jongy) [inline gprofiler/merge.py:373]: Metadata is not returned?

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:682]: In one site you use `args.upload_results or args.subcommands == "upload-file"` and in this site you use `args.upload_results or args.file_path`. Let's be consistent. You can create a helper function that gets the `args` object and returns true/false if we should be uploading.

Alternatively - and possible simpler - is to add after the `parse_args` call:
```
if args.subcommand == "upload-file":
   args.upload_results = True
```
then we don't get the 2 branches of the code.

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:533]:
```suggestion
        help="Path for the collapsed file to be uploaded",
```

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:532]: The default of `default` is None, I think :) No need to specify

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:530]: That's the default for `action`, no need to specify?

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:409]: Typing please

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:419]:
```suggestion
    if args.collect_metadata:
        metadata = get_current_metadata(get_static_metadata(spawn_time=spawn_time, run_args=args.__dict__)))
    else:
        metadata = {"hostname": get_hostname()}
```
simpler

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:381]: Don't need `total_samples`

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:327]: `self._gpid` is saved across sessions - you can return it from `_submit_profile_logged` and write it back to `self._gpid` here.

File: gprofiler/main.py
Comment (Jongy) [reply gprofiler/main.py:387]: See https://github.com/Granulate/gprofiler/pull/546/files#r1008565473

File: gprofiler/merge.py
Comment (Jongy) [reply gprofiler/merge.py:373]: Or actually, it should be embedded back into the first profile line. It's currently not done

**Comment by Jongy** [comment]
If I run with no args, I get:
```
root@40f149145ccb:/app# python3 -m gprofiler
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/app/gprofiler/__main__.py", line 8, in <module>
    main()
  File "/app/gprofiler/main.py", line 799, in main
    args = parse_cmd_args()
  File "/app/gprofiler/main.py", line 682, in parse_cmd_args
    if args.upload_results or args.file_path:
AttributeError: 'Namespace' object has no attribute 'file_path'
```

Let's ensure the "standard" behavior is kept in that case, i.e the same error message that we used to get before `upload-file`.

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:535]: Cool. Liked this solution :)

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:428]: Assert is the same condition as the `if`, so it will never be triggered. I suppose the `if` is meant to be `local_start_time is None or local_end_time is None`? (notice the `and` -> `or` diff)

File: gprofiler/main.py
Comment (Jongy) [reply gprofiler/main.py]: Good now :)

**Review by Jongy — CHANGES_REQUESTED** [review]
Finished reviewing, got some comments, overall looks good - that's the way this API & UX should be.

File: gprofiler/main.py
Comment (pfilipko1) [reply gprofiler/main.py:894]: Are we sure there will be no other subcommands in the future?

Comment (Jongy) [reply gprofiler/main.py:894]: Once we add them, the assert will trigger and we will remove it :)

**Comment by Jongy** [comment]
Many lint errors, please fix

**Comment by Jongy** [comment]
Also, the main repo's master appears a few commits ahead from your fork's master, you should pull the main fork master before merging "master" (otherwise you're not as up to date as could be)

File: README.md
Comment (Jongy) [inline README.md:123]: Mention that it's only for docker mode (maybe you meant "in case of docker run, ..."?)

File: gprofiler/utils/__init__.py
Comment (Jongy) [inline gprofiler/utils/__init__.py:361]:
```suggestion
    return datetime.datetime.fromisoformat(time)
```

File: gprofiler/merge.py
Comment (Jongy) [inline gprofiler/merge.py:361]: I suppose you're casing it here to replace the "function" object with a string?

In any case, we don't need it, so you can just ensure `func` is popped out from `run_arguments` (we pass `args.__dict__` to `GProfiler` class, you can pass a copy of the dict with `func` popped). Please explain in comment why it's done, because this code isn't clear

File: gprofiler/merge.py
Comment (Jongy) [inline gprofiler/merge.py:352]: Ensure the line starts with `#` as the format expects. (can be `assert line.startswith(...)`)

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:526]: Make it a required argument?

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:892]: This runs after `verify_preconditions()`, so we still require root & init PID NS. Let's ensure these are not requried, I think best would be to refrain from calling `verify_preconditions()` if we're in `upload-file` mode.

**Review by Jongy — CHANGES_REQUESTED** [review]
I managed to try it, works fine :) I have no more UX comments - `./gprofiler` behaves good when passing `--help`, `upload-file --help` and no arguments at all. I also managed to upload a file. We can close the existing comments and we're done

File: gprofiler/merge.py
Comment (pfilipko1) [reply gprofiler/merge.py:365]: You are right, the purpose was to omit json errors when function object was encountered. I'll pop it from run arguments then.

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:406]: If an exception was raised - `response_dict` is not assigned and this will result in an exception.

I think this code can be placed in `else` clause of the `try`, and at the end of the function, as a fallback, you can use the default value of `""`.

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:805]:
```suggestion
    if args.subcommand != "upload-file":
```

Let's be conservative, and be explicit for that case.

**Review by Jongy — APPROVED** [review]

**Pull request merged**
